lm_human_preference_details/train_reward_jax.py

```diff
@@ -0,0 +1,765 @@
+            reward_head=reward_head,
+        ),
+        axis_name="batch",
+        donate_argnums=(0,),
+    )
+    p_val_step = jax.pmap(
+        functools.partial(
+            val_step,
+            args=args,
+            reward_backbone=reward_backbone,
+            reward_head=reward_head,
+        ),
+        axis_name="batch",
+    )
+
+    normalization_dataset = MyDataset(
+        DATASET[args.task.query_dataset],
+        tokenizer,
+        args.task.query_length,
+        seed=args.seed,
+        start_text=args.task.start_text,
+        end_text=args.task.end_text,
+    )
+    normalization_dataloader = DataLoader(
+        normalization_dataset, batch_size=args.rollout_batch_size
+    )
+    iter_normalization_dataloader = iter(normalization_dataloader)
+
+    generation_config = transformers.GenerationConfig(
+        max_new_tokens=args.task.response_length,
+        min_new_tokens=args.task.response_length,
+        temperature=args.task.temperature,
+        top_k=0.0,
+        top_p=1.0,
+        do_sample=True,
+        pad_token_id=args.pad_token_id,
+    )
+
+    if args.normalize_before:
+        print("===Normalize reward model *before* training===")
+
+        # pylint: disable=E1101:no-member
+        print(
+            "before normalization. "
+            + f"Gain: {reward_state.params.head_params['params']['reward_gain']}"
+            + f" Bias: {reward_state.params.head_params['params']['reward_bias']}"
+        )
+
+        reward_state = normalize(
+            args,
+            tokenizer,
+            untrained_model,
+            reward_state,
+            iter_normalization_dataloader,
+            generation_config,
+            reward_backbone,
+            reward_head,
+        )
+
+        print(
+            "after normalization. "
+            + f"Gain: {reward_state.params.head_params['params']['reward_gain']}"
+            + f" Bias: {reward_state.params.head_params['params']['reward_bias']}"
+        )
+
+    reward_state = jax_utils.replicate(reward_state)
+
+    # `labeled_dataset` has keys
+    # `['sample0', 'query', 'best', 'sample3', 'sample1', 'sample2']`
+    labeled_dataset = load_dataset(
+        "vwxyzjn/lm-human-preferences", data_files=[args.label_dataset],
+    )["train"]
+    print("Num labels found in source:", len(labeled_dataset))
+    print("training on", args.labels.num_train, "in batches of", args.local_batch_size)
+
+    all_queries_responses, all_labels = prepare_left_padded_query_responses_with_labels(
+        labeled_dataset, args
+    )
+
+    assert args.labels.num_train < all_queries_responses.shape[0]
+    train_queries_responses = all_queries_responses[: args.labels.num_train]
+    train_labels = all_labels[: args.labels.num_train]
+
+    val_queries_responses = all_queries_responses[args.labels.num_train :]
+    val_labels = all_labels[args.labels.num_train :]
+
+    train_iter = get_dataloader_iter(
+        jax.random.PRNGKey(args.seed),
```

Comment:
The common practice seems to be creating an RNG then splitting them in the beginning. E.g., 

```
    key = jax.random.PRNGKey(args.seed)
    key, network_key, actor_key, critic_key = jax.random.split(key, 4)
```